manager/controlapi/service.go

```diff
@@ -102,6 +103,29 @@ func validateUpdate(uc *api.UpdateConfig) error {
 	return nil
 }
 
+func validateContainerSpec(container *api.ContainerSpec) error {
+	if container == nil {
+		return grpc.Errorf(codes.InvalidArgument, "ContainerSpec: missing in service spec")
+	}
+
+	if container.Hostname != "" {
+		// Validate if the given hostname is RFC 1123
+		matched, _ := regexp.MatchString("^(([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])\\.)*([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])$", container.Hostname)
```

Comment:
A few things:

- make a function for this
- pre-compile the regexp (use the global pattern with regexp.MustCompile)